fix(ServiceExport): Reject Conflicting Alias Names for ServiceExport object (#315)

soharab-ic opened the pull request:
<!--
Type of change: Title of the PR should clearly mention which type of PR is this, you can select any of the below-mentioned types:

- docs() - The PR contains Documentation ONLY changes. 
- feat() - The PR contains new feature/enhancements.
- fix() - The PR contains a bug fix.

Example Title: 
feat(): New field addition for Cluster CRD 
-->

# Description
<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
--> The validating webhook will validate the ServiceExport object and admission of ServiceExport objects with conflicting alias names will be rejected.

Fixes  <!-- Mention any issues which might be fixed on this PR merge. --> 315

## How Has This Been Tested?
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

test-cases
- [ ] Test case A
- [ ] Test case B
-->

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have ran `go fmt`
* [ ] I have updated the helm chart as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [x] I have added all the required unit test cases.
* [ ] I have verified the E2E test cases with new code changes.
* [ ] I have added all the required E2E test cases.

## Does this PR introduce a breaking change?
<!--
If NO, leave the release-note block blank.
If YES, a release note is required:
Enter your extended release note in the block below. If the PR requires additional manual action from users switching to the new version, include the string "action-required".
-->

```release-note

```

YachikaRalhan commented on pkg/webhook/pod/webhook.go:
```diff
@@ -261,6 +281,33 @@ func MutateDaemonSet(ds *appsv1.DaemonSet, sliceName string) *appsv1.DaemonSet {
+
+	for _, serviceExport := range serviceExportList.Items {
+		// In case we are updating an existing ServiceExport resource
+		if svcex.ObjectMeta.Name == serviceExport.ObjectMeta.Name &&
```
Why are you skipping the existing service export updation validation?

soharab-ic replied:
If we are updating a serviceexport object svcex-1:
1. serviceExportList will contain svcex-1.
2. Comparing the aliases of updated svcex-1 and existing svcex-1 may give us conflict.
3. Even if we are updating the aliases, it is safe to compare with aliases of other serviceexport objects not the same seviceexport object.

rahulii commented:
Looks good overall, please test it once !
Also please add unit test cases!

soharab-ic commented:
The issue 338 impacts this PR. Will complete the testing once it is resolved.

soharab-ic commented:
> Looks good overall, please test it once ! Also please add unit test cases!

Already added unit test cases.
https://github.com/infracloudio/kubeslice-worker-operator/pull/2/files#diff-45a421353c537e3010114d2de28717519d2e86df663d7130d7cc64bca4f8e48cR158

soharab-ic committed:
fix(ServiceExport): Reject Conflicting Alias Names for ServiceExport object

Added a validating webhook to verify alias names in serviceexport object
and reject and if there are conflicting aliases

Fixes #315

Signed-off-by: Md Soharab Ansari <soharab.ansari@infracloud.io>
